src/reader/buffered_reader.rs

```diff
@@ -393,8 +396,14 @@ impl<R: BufRead> Reader<R> {
             buf.clear();
         }))
     }
+
+    /// Same as `read_to_end_into` but returns all the bytes read instead of the indices.
+    pub fn read_to_end_to_string(&mut self, end: QName, buf: &mut Vec<u8>) -> Result<Vec<u8>> {
```

Comment:
That function should be named `read_text_into` and have signature
```rust
pub fn read_text_into<'b>(&mut self, end: QName, buf: &'b mut Vec<u8>) -> Result<Cow<'b, str>>
```

I also expect the same set of tests as for `read_text` and implementation of `read_text_into_async`.

Both should have the similar documentation as `read_text` has. You can just copy existing documentation of `read_text` and edit the examples section. Also look at the differences in documentation of other paired functions and write the doc the same way.

Reply:
OK. I think we actually need two buffers. One buffer is for temporary storage of event data, and one buffer is to store the actual output string that will be returned to the user. 

Also I'm not sure why we would need a `Cow<str>` as we already have a buffer we can use to put the str in. The return type should be `Result<&'b str>`

So the signature will be:
```rust
pub fn read_text_into<'b>(&mut self, end: QName, event_buf: &mut Vec<u8>, output_buf: &'b mut Vec<u8>) -> Result<&'b str>
```

Reply:
`Cow` is required because it is return type of the `decode` method. You cannot use `from_utf8` to get a string. When feature `encoding` is activated, we can parse documents in almost all encodings supported by `encoding_rs` and we should decode bytes to string using XML encoding.

> One buffer is for temporary storage of event data

It is better to avoid this. That likely will require more work that just reuse existing parsing methods, but `output_buf` will contain all needed data for the events. This is why we need separate methods for buffered and borrowing readers and one way to implement that -- add a new method to `XmlSource`. But you're right, this is not necessarily required. You can try to find another way, the way how `async` is supported probably can give you some ideas